File: api-artist-workspace/src/main/java/org/example/workspace/security/CustomUserDetails.java
package org.example.workspace.security;

import lombok.*;
import org.example.workspace.entity.User;
import org.example.workspace.entity.code.RoleType;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.List;

@Builder(access = AccessLevel.PRIVATE)
@NoArgsConstructor(access = AccessLevel.PRIVATE)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class CustomUserDetails implements UserDetails {
    private User user;
    private RoleType roleType;
    @Getter
    private Long id;

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return List.of(new SimpleGrantedAuthority(roleType.name()));
    }

    @Override
    public String getPassword() {
        return user.getPassword();
    }

    @Override
    public String getUsername() {
        return user.getLoginId();
    }

    @Override
    public boolean isAccountNonExpired() {
        return UserDetails.super.isAccountNonExpired();
    }

    @Override
    public boolean isAccountNonLocked() {
        return UserDetails.super.isAccountNonLocked();
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return UserDetails.super.isCredentialsNonExpired();
    }

    @Override
    public boolean isEnabled() {
        return user.getIsActivated();
    }

    public static CustomUserDetails create(User user, RoleType role) {
        return CustomUserDetails.builder()
                .id(user.getId())
                .user(user)
                .roleType(role)
                .build();
    }

    public RoleType getRoleType() {
        String roleName = getAuthorities().stream()
                .findFirst()
                .map(GrantedAuthority::getAuthority)
                .orElseThrow(() -> new AccessDeniedException("Access Denied: User has no role"));
        return RoleType.valueOf(roleName);
    }
}

File: level-up-backend-gpt/level2-book-management-system/src/main/kotlin/org/example/level2bookmanagementsystem/security/core/CustomUserDetails.kt
package org.example.level2bookmanagementsystem.security.core

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails

class CustomUserDetails(
    private val user: User
): UserDetails {

    // 사용자의 권한 목록을 반환
    override fun getAuthorities(): Collection<GrantedAuthority> {
        return user.authorities
    }

    // 사용자의 비밀번호를 반환
    override fun getPassword(): String {
        return user.password
    }

    // 사용자의 이름(아이디)을 반환
    override fun getUsername(): String {
        return user.username
    }

    // 계정이 만료되지 않았는지 여부 반환
    override fun isAccountNonExpired(): Boolean {
        return super.isAccountNonExpired()
    }

    // 계정이 잠겨있지 않은지 여부 반환
    override fun isAccountNonLocked(): Boolean {
        return super.isAccountNonLocked()
    }

    // 자격 증명이 만료되지 않았는지 여부 반환
    override fun isCredentialsNonExpired(): Boolean {
        return super.isCredentialsNonExpired()
    }

    // 계정이 활성화되어 있는지 여부 반환
    override fun isEnabled(): Boolean {
        return super.isEnabled()
    }
}